src/query/storages/parquet/src/parquet_rs/meta.rs

```diff
@@ -142,14 +189,44 @@ pub async fn read_parquet_metas_batch(
         }));
     }
 
-    // TODO(parquet): how to limit the memory when running this method is to be determined.
+    check_memory_usage(max_memory_usage)?;
+    Ok(metas)
+}
+
+pub async fn read_parquet_metas_batch_for_copy(
+    file_infos: Vec<(String, u64)>,
+    op: Operator,
+    max_memory_usage: u64,
+) -> Result<Vec<Arc<FullParquetMeta>>> {
+    let mut metas = Vec::with_capacity(file_infos.len());
+    for (location, size) in file_infos {
+        let meta = Arc::new(
+            databend_common_storage::parquet_rs::read_metadata_async(&location, &op, Some(size))
+                .await?,
+        );
+        if unlikely(meta.file_metadata().num_rows() == 0) {
+            // Don't collect empty files
+            continue;
+        }
+        metas.push(Arc::new(FullParquetMeta {
+            location,
+            size,
+            meta,
+            row_group_level_stats: None,
+        }));
+    }
+    check_memory_usage(max_memory_usage)?;
+    Ok(metas)
+}
+
+// TODO(parquet): how to limit the memory when running this method is to be determined.
+fn check_memory_usage(max_memory_usage: u64) -> Result<()> {
```

Comment:
This may be useless, @BohuTANG just told me and after I tried, I found it true.

Reply:
I think no need to check this, it is duplicate with the soft oom in the global runtime tracker:
https://github.com/datafuselabs/databend/blob/ecf1c0df3d930d1c140356965f25409e2a2957c9/src/common/base/src/runtime/runtime_tracker.rs#L155-L165

Reply:
Oh, please ignore my comments:
`if max_memory_usage as i64 - used < 100 * 1024 * 1024` is more strict than the global checker.